File: Cores/OSOL.Extremum.Core.DotNet/CodeFeatures/FunctionParser/Trees/TreeD.cs
using System;

namespace OSOL.Extremum.Core.DotNet.CodeFeatures.FunctionParser.Trees
{
    public static class TreeD
    {
        public class AdditionTree : BinaryOpTree<double>
        {
            public AdditionTree(Tree<double> left, Tree<double> right)
            {
                this.Left = left;
                this.Right = right;
                this.Op = (x, y) => x + y;
            }
        }
        
        public class SubtractionTree : BinaryOpTree<double>
        {
            public SubtractionTree(Tree<double> left, Tree<double> right)
            {
                this.Left = left;
                this.Right = right;
                this.Op = (x, y) => x - y;
            }
        }
        
        public class MultiplicationTree : BinaryOpTree<double>
        {
            public MultiplicationTree(Tree<double> left, Tree<double> right)
            {
                this.Left = left;
                this.Right = right;
                this.Op = (x, y) => x * y;
            }
        }
        
        public class DivisionTree : BinaryOpTree<double>
        {
            public DivisionTree(Tree<double> left, Tree<double> right)
            {
                this.Left = left;
                this.Right = right;
                this.Op = (x, y) => x / y;
            }
        }
        
        public class PowerTree : BinaryOpTree<double>
        {
            public PowerTree(Tree<double> left, Tree<double> right)
            {
                this.Left = left;
                this.Right = right;
                this.Op = Math.Pow;
            }
        }

        public class NegTree : UnaryOpTree<double>
        {
            public NegTree(Tree<double> subTree)
            {
                this.SubTree = subTree;
                this.Op = x => -x;
            }
        }

        public class SinTree : UnaryOpTree<double>
        {
            public SinTree(Tree<double> subTree)
            {
                this.SubTree = subTree;
                this.Op = Math.Sin;
            }
        }

        public class CosTree : UnaryOpTree<double>
        {
            public CosTree(Tree<double> subTree)
            {
                this.SubTree = subTree;
                this.Op = Math.Cos;
            }
        }

        public class ExpTree : UnaryOpTree<double>
        {
            public ExpTree(Tree<double> subTree)
            {
                this.SubTree = subTree;
                this.Op = Math.Exp;
            }
        }

        public class AbsTree : UnaryOpTree<double>
        {
            public AbsTree(Tree<double> subTree)
            {
                this.SubTree = subTree;
                this.Op = Math.Abs;
            }
        }

        public class LnTree : UnaryOpTree<double>
        {
            public LnTree(Tree<double> subTree)
            {
                this.SubTree = subTree;
                this.Op = Math.Log;
            }
        }

        public class SqrtTree : UnaryOpTree<double>
        {
            public SqrtTree(Tree<double> subTree)
            {
                this.SubTree = subTree;
                this.Op = Math.Sqrt;
            }
        }
    }
}

File: Cores/ExtremumScalaCore/src/main/scala/OSOL/Extremum/Core/Scala/CodeFeatures/FunctionParser/Trees/TreeD.scala
package OSOL.Extremum.Core.Scala.CodeFeatures.FunctionParser.Trees

import OSOL.Extremum.Core.Scala.CodeFeatures.FunctionParser.Trees.Tree._
import OSOL.Extremum.Core.Scala.Vectors.VectorObject

object TreeD {

  class AdditionTree(left: Tree[Double], right: Tree[Double]) extends BinaryOpTree[Double](left, right, (x, y) => x + y)

  class SubtractionTree(left: Tree[Double], right: Tree[Double]) extends BinaryOpTree[Double](left, right, (x, y) => x - y)

  class MultiplicationTree(left: Tree[Double], right: Tree[Double]) extends BinaryOpTree[Double](left, right, (x, y) => x * y)

  class DivisionTree(left: Tree[Double], right: Tree[Double]) extends BinaryOpTree[Double](left, right, (x, y) => x / y)

  class PowerTree(left: Tree[Double], right: Tree[Double]) extends BinaryOpTree[Double](left, right, (x, y) => math.pow(x, y))

  class NegTree(subTree: Tree[Double]) extends UnaryOpTree[Double](subTree, x => -x)

  class SinTree(subTree: Tree[Double]) extends UnaryOpTree[Double](subTree, x => math.sin(x))

  class CosTree(subTree: Tree[Double]) extends UnaryOpTree[Double](subTree, x => math.cos(x))

  class ExpTree(subTree: Tree[Double]) extends UnaryOpTree[Double](subTree, x => math.exp(x))

  class AbsTree(subTree: Tree[Double]) extends UnaryOpTree[Double](subTree, x => math.abs(x))

  class LnTree(subTree: Tree[Double]) extends UnaryOpTree[Double](subTree, x => math.log(x))

  class SqrtTree(subTree: Tree[Double]) extends UnaryOpTree[Double](subTree, x => math.sqrt(x))

  class ConstantTree(value: Double) extends Tree[Double] {
    final override def calculate(v: VectorObject[Double]): Double = value
  }

  class VariableTree(varName: String) extends Tree[Double] {
    final override def calculate(v: VectorObject[Double]): Double = v(varName)
  }

}